File: dqops/src/main/java/com/dqops/core/dqocloud/users/DqoUserRolesModel.java
/*
 * Copyright © 2021-Present DQOps, Documati sp. z o.o. (support@dqops.com)
 *
 * This file is licensed under the Business Source License 1.1,
 * which can be found in the root directory of this repository.
 *
 * Change Date: This file will be licensed under the Apache License, Version 2.0,
 * four (4) years from its last modification date.
 */

package com.dqops.core.dqocloud.users;

import com.dqops.cloud.rest.model.DqoUserModel;
import com.dqops.core.dqocloud.login.DqoUserRole;
import com.dqops.core.principal.UserDomainIdentity;
import com.dqops.metadata.settings.domains.LocalDataDomainSpec;
import com.dqops.utils.docs.generators.SampleStringsRegistry;
import com.dqops.utils.docs.generators.SampleValueFactory;
import com.fasterxml.jackson.annotation.JsonPropertyDescription;
import io.swagger.annotations.ApiModel;
import lombok.Data;

import java.util.*;

/**
 * DQOps user model - identifies a user in a multi-user DQOps deployment and the user's roles.
 */
@ApiModel(value = "DqoUserRolesModel", description = "DQOps user model - identifies a user in a multi-user DQOps deployment and the user's roles.")
@Data
public class DqoUserRolesModel {
    /**
     * User's email that identifies the user.
     */
    @JsonPropertyDescription("User's email that identifies the user.")
    private String email;

    /**
     * Account role.
     */
    @JsonPropertyDescription("User role at the whole account level. This role is applicable to all data domains.")
    private DqoUserRole accountRole;

    /**
     * User roles within each data domain. Data domains are supported in an ENTERPRISE version of DQOps and they are managed by the SaaS components of DQOps Cloud.
     */
    @JsonPropertyDescription("User roles within each data domain. Data domains are supported in an ENTERPRISE version of DQOps and they are managed by the SaaS components of DQOps Cloud.")
    private Map<String, DqoUserRole> dataDomainRoles = new LinkedHashMap<>();

    /**
     * Converts the local user model to the cloud user model.
     * @return Cloud user model.
     */
    public DqoUserModel toCloudUserModel() {
        DqoUserModel dqoCloudUserModel = new DqoUserModel();
        dqoCloudUserModel.setEmail(this.getEmail());
        dqoCloudUserModel.setAccountRole(this.getAccountRole().convertToApiAccountRoleEnum());

        if (this.getDataDomainRoles() != null) {
            LinkedHashMap<String, DqoUserModel.InnerEnum> cloudDataDomainRoles = new LinkedHashMap<>();
            for (Map.Entry<String, DqoUserRole> localDomainRolePair : this.getDataDomainRoles().entrySet()) {
                if (localDomainRolePair.getValue() != DqoUserRole.NONE) {
                    String domainName = localDomainRolePair.getKey();
                    if (Objects.equals(domainName, UserDomainIdentity.ROOT_DOMAIN_ALTERNATE_NAME)) {
                        domainName = UserDomainIdentity.ROOT_DATA_DOMAIN;
                    }
                    cloudDataDomainRoles.put(domainName, localDomainRolePair.getValue().convertToApiDomainRoleEnum());
                }
            }

            dqoCloudUserModel.setDataDomainRoles(cloudDataDomainRoles);
        }

        return dqoCloudUserModel;
    }

    /**
     * Creates a local user model from a cloud user model.
     * @param cloudUserModel Cloud (data provider) user model.
     * @param allDataDomains All data domains, even those that the user has no rights. Pass null when the instance is not supporting data domains.
     * @return Local user model.
     */
    public static DqoUserRolesModel createFromCloudModel(DqoUserModel cloudUserModel, Collection<LocalDataDomainSpec> allDataDomains) {
        DqoUserRolesModel localUserModel = new DqoUserRolesModel() {{
            setEmail(cloudUserModel.getEmail());
            setAccountRole(DqoUserRole.convertFromApiAccountRoleEnum(cloudUserModel.getAccountRole()));
        }};

        if (allDataDomains != null) {
            LinkedHashMap<String, DqoUserRole> localDataDomains = new LinkedHashMap<>();
            if (cloudUserModel.getDataDomainRoles() != null) {
                for (Map.Entry<String, DqoUserModel.InnerEnum> cloudDomainRolePair : cloudUserModel.getDataDomainRoles().entrySet()) {
                    String cloudDomainName = cloudDomainRolePair.getKey();
                    String displaySafeDataDomainName = Objects.equals(cloudDomainName, UserDomainIdentity.ROOT_DATA_DOMAIN) ?
                            UserDomainIdentity.ROOT_DOMAIN_ALTERNATE_NAME : cloudDomainName;

                    localDataDomains.put(displaySafeDataDomainName, DqoUserRole.convertFromApiDomainRole(cloudDomainRolePair.getValue()));
                }
            }

            for (LocalDataDomainSpec localDataDomainSpec : allDataDomains) {
                String localDomainName = localDataDomainSpec.getDataDomainName();

                if (cloudUserModel.getDataDomainRoles() != null && cloudUserModel.getDataDomainRoles().containsKey(
                        Objects.equals(localDomainName, UserDomainIdentity.ROOT_DOMAIN_ALTERNATE_NAME) ? UserDomainIdentity.ROOT_DATA_DOMAIN : localDomainName)) {
                    continue; // already mapped
                }

                // add missing data domains with no access rights
                String displaySafeDataDomainName = Objects.equals(localDomainName, UserDomainIdentity.ROOT_DATA_DOMAIN) ?
                        UserDomainIdentity.ROOT_DOMAIN_ALTERNATE_NAME : localDomainName;

                localDataDomains.put(displaySafeDataDomainName, DqoUserRole.NONE);
            }

            localUserModel.setDataDomainRoles(localDataDomains);
        }

        return localUserModel;
    }

    /**
     * Documentation sample factory.
     */
    public static class DqoUserRolesModelSampleFactory implements SampleValueFactory<DqoUserRolesModel> {
        @Override
        public DqoUserRolesModel createSample() {
            return new DqoUserRolesModel() {{
                setEmail(SampleStringsRegistry.getEmail());
                setAccountRole(DqoUserRole.OPERATOR);
                setDataDomainRoles(new LinkedHashMap<>() {{
                    put(UserDomainIdentity.ROOT_DOMAIN_ALTERNATE_NAME, DqoUserRole.EDITOR);
                    put("datalake", DqoUserRole.EDITOR);
                }});
            }};
        }
    }
}

File: distribution/python/dqops/client/models/dqo_user_roles_model.py
from typing import TYPE_CHECKING, Any, Dict, List, Type, TypeVar, Union

from attrs import define as _attrs_define
from attrs import field as _attrs_field

from ..models.dqo_user_role import DqoUserRole
from ..types import UNSET, Unset

if TYPE_CHECKING:
    from ..models.dqo_user_roles_model_data_domain_roles import (
        DqoUserRolesModelDataDomainRoles,
    )


T = TypeVar("T", bound="DqoUserRolesModel")


@_attrs_define
class DqoUserRolesModel:
    """DQOps user model - identifies a user in a multi-user DQOps deployment and the user's roles.

    Attributes:
        email (Union[Unset, str]): User's email that identifies the user.
        account_role (Union[Unset, DqoUserRole]):
        data_domain_roles (Union[Unset, DqoUserRolesModelDataDomainRoles]): User roles within each data domain. Data
            domains are supported in an ENTERPRISE version of DQOps and they are managed by the SaaS components of DQOps
            Cloud.
    """

    email: Union[Unset, str] = UNSET
    account_role: Union[Unset, DqoUserRole] = UNSET
    data_domain_roles: Union[Unset, "DqoUserRolesModelDataDomainRoles"] = UNSET
    additional_properties: Dict[str, Any] = _attrs_field(init=False, factory=dict)

    def to_dict(self) -> Dict[str, Any]:
        email = self.email
        account_role: Union[Unset, str] = UNSET
        if not isinstance(self.account_role, Unset):
            account_role = self.account_role.value

        data_domain_roles: Union[Unset, Dict[str, Any]] = UNSET
        if not isinstance(self.data_domain_roles, Unset):
            data_domain_roles = self.data_domain_roles.to_dict()

        field_dict: Dict[str, Any] = {}
        field_dict.update(self.additional_properties)
        field_dict.update({})
        if email is not UNSET:
            field_dict["email"] = email
        if account_role is not UNSET:
            field_dict["accountRole"] = account_role
        if data_domain_roles is not UNSET:
            field_dict["dataDomainRoles"] = data_domain_roles

        return field_dict

    @classmethod
    def from_dict(cls: Type[T], src_dict: Dict[str, Any]) -> T:
        from ..models.dqo_user_roles_model_data_domain_roles import (
            DqoUserRolesModelDataDomainRoles,
        )

        d = src_dict.copy()
        email = d.pop("email", UNSET)

        _account_role = d.pop("accountRole", UNSET)
        account_role: Union[Unset, DqoUserRole]
        if isinstance(_account_role, Unset):
            account_role = UNSET
        else:
            account_role = DqoUserRole(_account_role)

        _data_domain_roles = d.pop("dataDomainRoles", UNSET)
        data_domain_roles: Union[Unset, DqoUserRolesModelDataDomainRoles]
        if isinstance(_data_domain_roles, Unset):
            data_domain_roles = UNSET
        else:
            data_domain_roles = DqoUserRolesModelDataDomainRoles.from_dict(
                _data_domain_roles
            )

        dqo_user_roles_model = cls(
            email=email,
            account_role=account_role,
            data_domain_roles=data_domain_roles,
        )

        dqo_user_roles_model.additional_properties = d
        return dqo_user_roles_model

    @property
    def additional_keys(self) -> List[str]:
        return list(self.additional_properties.keys())

    def __getitem__(self, key: str) -> Any:
        return self.additional_properties[key]

    def __setitem__(self, key: str, value: Any) -> None:
        self.additional_properties[key] = value

    def __delitem__(self, key: str) -> None:
        del self.additional_properties[key]

    def __contains__(self, key: str) -> bool:
        return key in self.additional_properties
